src/modm/platform/extint/sam/extint.cpp

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2020, Erik Henriksson
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include "extint.hpp"
+
+void isr();
+
+namespace modm
+{
+namespace platform
+{
+
+void
+ExternalInterrupt::initialize(ClockGenerator clockGen) {
+	NVIC_DisableIRQ(EIC_IRQn);
+	NVIC_ClearPendingIRQ(EIC_IRQn);
+	NVIC_SetPriority(EIC_IRQn, 0);
+	NVIC_EnableIRQ(EIC_IRQn);
+
+	GenericClockController::enable<Peripheral::Eic>(clockGen);
+
+	// Enable EIC
+	EIC->CTRL.bit.ENABLE = 1;
+	int waitCycles = 2048;
+	while (EIC->STATUS.bit.SYNCBUSY == 1 && --waitCycles);
+}
+
+std::function<void()> ExternalInterrupt::handlers_[16] = {};
+
+// FIXME: Figure out if it is worth using function pointers here instead
+//        to get rid of std::function overhead.
+MODM_ISR(EIC) {
+	uint32_t int_flag = EIC->INTFLAG.reg;
```

Comment:
Hm, you could also do it differently: by having a list of tuples of (instance, handler) and then run through that in the ISR. That way you can also trivially add multiple handlers per instance. You could make the list length an lbuild option.

Currently you'll end up with a sparse handler array, since it's unlikely you'll attach a handler to every instance, and you're still limited to only one handler even though the memory for the other handers is unused.

Reply:
Cool, I did not know of that intrinsic!

Now, we should be down to 28 instructions overall overhead and 15 instructions latency overhead on Cortex-M0+ for the case of a single interrupt flag set (which I guess is the case to optimize for). Unfortunately doing your tuple suggestion totally breaks this, as I cannot anymore efficiently index into the handlers list but need to resort to scanning it (which means way more latency overhead).

Suggestion to address this thread is available here: https://godbolt.org/z/839W4T

Sparse array vs list is a RAM vs interrupt latency trade-off, and this is a small enough list that hashmap won't overperform a sparse array. I believe latency is more important than memory usage here (it is already down to sizeof(void*) * number of interrupt lines), but I can see how the ATTiny folks would disagree with me.

I will do these changes after hearing your feedback on that.

Reply:
I think the latency advantages of the CLZ implementation significantly outweigh any flexibility you get from a list, especially since you can still group your handlers manually as you pointed out above. So let's go with your godbolt code.